find_tail_calls.py

```diff
@@ -0,0 +1,75 @@
+from dataclasses import dataclass, field
+from typing import Any, List, Optional, Set, cast
+
+import sexp
+from visitor import Visitor
+
+
+@dataclass
+class TailCallData:
+    call: sexp.SCall
+    func_params: List[sexp.SSym] = field(default_factory=list)
+
+    def __hash__(self) -> int:
+        return hash(id(self.call))
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, TailCallData):
+            return False
+        return hash(self) == hash(other)
+
+
+class TailCallFinder(Visitor):
+    def __init__(self) -> None:
+        self._tail_calls: List[TailCallData] = []
+
+        self._current_function: Optional[sexp.SFunction] = None
+
+    @property
+    def tail_calls(self) -> List[TailCallData]:
+        return list(self._tail_calls)
+
+    def visit_SFunction(self, func: sexp.SFunction) -> None:
+        if func.is_lambda:
+            # We're only looking for statically-detectable tail recursion.
+            return
+
+        body_exprs = list(func.body)
+        assert len(body_exprs) > 0, 'Function bodies must not be empty'
+
+        for body_expr in body_exprs[:-1]:
```

Comment:
This is bailing out if there are any non-tail recursive calls, right?

I believe:
```scheme
(define (M n)
  (if (> n 100)
     (- n 10)
     (M (M (+ n 11)))))
```
can be safely optimized to
```
M(n) =
  while n <= 100
    n := M(n + 11)
  n - 10
```